velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -57,6 +90,82 @@ void ensureRegexIsCompatible(
   }
 }
 
+// Blocks patterns that contain character class union, intersection, or
+// difference because these are not understood by RE2 and will be parsed as a
+// different pattern than in java.util.regex.
+void ensureRegexIsCompatible(
+    const char* functionName,
+    const VectorPtr& patternVector) {
+  if (!patternVector ||
+      patternVector->encoding() != VectorEncoding::Simple::CONSTANT) {
+    VELOX_USER_FAIL("{} requires a constant pattern.", functionName);
+  }
+  if (patternVector->isNullAt(0)) {
+    return;
+  }
+  const StringView pattern =
+      patternVector->as<ConstantVector<StringView>>()->valueAt(0);
+  // Call the factored out function to check the pattern.
+  checkForCompatiblePattern(
+      std::string(pattern.data(), pattern.size()), functionName);
+}
+
+/// Full implementation of RegexReplace found in SparkSQL only,
+/// due to semantic mismatches betweeen Spark and Presto
+/// regex_replace(string, pattern, overwrite) → string
+/// regex_replace(string, pattern, overwrite, position) → string
+///
+/// If a string has a substring that matches the given pattern, replace
+/// the match in the string wither overwrite and return the string. If
+/// optional paramter position is provided, only make replacements
+/// after that positon in the string (1 indexed).
+///
+/// If position <= 0, throw error.
+/// If position > length string, return string.
+template <typename T>
+struct regexReplaceFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& string,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace) {
+    checkForCompatiblePattern(pattern);
+    re2::RE2 patternRE(toStringPiece(pattern));
+    re2::StringPiece replaceSP = toStringPiece(replace);
+    std::string stringS = std::string(string.data(), string.size());
+    int replacements = RE2::GlobalReplace(&stringS, patternRE, replaceSP);
+    result.resize(stringS.size());
+    std::memcpy(result.data(), stringS.data(), stringS.size());
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& string,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    VELOX_CHECK(!(position - 1 < 0));
+    checkForCompatiblePattern(pattern);
+    re2::RE2 patternRE(toStringPiece(pattern));
+    re2::StringPiece replaceSP = toStringPiece(replace);
+    std::string stringS = std::string(string.data(), string.size());
```

Comment:
This code seems to be creating lots of intermediate strings. Is this necessary?

Reply:
The problem we are running into here is that the API for RE2:GlobalReplace is very rigid. It requires a std::string*, does not allow for a position parameter, and requires a StringPiece for replacement.  I made some very small changes to this code in my next push that offer slight optimizations (less copying and more StringPiece leveraging), but to the extent of my knowledge, there is no way to avoid intermediate strings with our specific use case.